### DISCUSSION TITLE
QSPI for ESP32

### DISCUSSION THREAD

**Discussion by obones** [RudolphRiedel/FT800-FT813#28] [Ideas]
Hello,

I'm about to receive the evaluation board for the BT817 and while reading its documentation, I see that there is a pin header with SPI/QSPI signals.
As I'm intending to drive it with an ESP32, I had a look at your examples, and as far as I can understand, it is using SPI to communicate, not QSPI. 
Do you think using QSPI would bring any benefit? And if yes, how would you suggest that I try investigating the required changes?

Regards

**Comment by RudolphRiedel** [comment]
Hi,

I have not tried to use QSPI.
The first reason is that most controllers I actively used do not have a QSPI interface.
The second reason is that I never had an issue with the data transfer rate of single SPI since I switched over to using DMA (and those controllers not supporting DMA are not supporting QSPI either).

And then this is application specific if the increased data rate would make a difference.
For example you could have the requirement to have the display up and running in under 300ms and the application needs to upload a high number of assets on startup. I never had such a requirement while the startup times were always pretty impressive. And loading assets is not really an issue anymore since we got BT81x with external flash.

Annother example would be playing video files but personally I consider this to be out of scope for EVE.
I may perhaps try some day to play a small video from flash.
But if had an application that would require playing large video files I would likely consider a different solution.

Next there may be a reason to drop the SPI clock. But before switching over to QSPI I would check how much of the 5ms I currently allow is actually needed and would either change the timing, even drop the framerate a little if required.
And there is always something to optimize in the display list to shave off a couple of bytes fro the SPI.

The last scenario I can think of right now is when you have one QSPI but more than one client.
For example SD-card and EVE.
But I am not even sure if this is a valid example, if it really is viable to have  more than one client on a QSPI.

To add QSPI would require two things.
First the initialisation needs to be changed in EVE_Init().
And then the SPI functions in EVE_target.c and EVE_target.h would require changes.

**Comment by davidjade** [comment]
FYI, the LVGL project has an ESP32 driver which can drive the EVE2 using QSPI at full speed using DMA. The QSPI/DMA driver code, which could be reused I suppose, is very ESP32 optimized for just bit-blt raw graphical data to the EVE, treating it as a dumb device (which is counter intuitive to say the least). It works and is even stable on breadboards with careful cabling at slightly more than official clock speed.

**Comment by RudolphRrr** [comment]
Yes, mis-using EVE as framebuffer would be annother use-case. :-)
So what actually is the use-case that potentially could benefit from using QSPI?

And regarding the ESP32, have a look here: https://github.com/RudolphRiedel/FT800-FT813/issues/14
The main issue with the ESP32 is that the supplied drivers are rather slow.
But in the end I got it work with DMA support.
For a normal use case with objects in a display list the only difference between using SPI and QSPI would be how long the SPI is busy sending after the buffer of upto 4kiB of objects has been prepared.

**Comment by obones** [comment]
Thanks, that clarifies the situation quite a lot, and I agree, using QSPI will most likely not be that useful to my project anyway.
Should it become necessary, I'll start with the pointers you gave me.
